{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import random\n",
    "random.seed(12)\n",
    "\n",
    "import spacy\n",
    "from spacy.tokenizer import Tokenizer\n",
    "sys.path.insert(1, '../../utils/')\n",
    "\n",
    "from bratUtils import myCorpus_brat2conll\n",
    "from conllUtils import describe_entities\n",
    "from myDocClass import ncbi_doc\n",
    "\n",
    "from collections import Counter\n",
    "%reload_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Training and dev corpus construction for NCBI-disease corpus\n",
    "\n",
    "We build a train and dev set from the [NCBI-disease corpus](https://www.ncbi.nlm.nih.gov/CBBresearch/Dogan/DISEASE/). This corpus contains 592 documents that we convert to conll files (suitable input format for the yaset tool)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "path2train = '../data/src/NCBItrainset_corpus.txt'\n",
    "path2train_write = '../data/ncbi_train'\n",
    "path2dev = '../data/src/NCBIdevelopset_corpus.txt'\n",
    "path2dev_write = '../data/ncbi_dev'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(path2train, 'r') as f:\n",
    "    train_data = f.read().splitlines()\n",
    "with open(path2dev, 'r') as f:\n",
    "    dev_data = f.read().splitlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "592"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_docs = []\n",
    "point = 0\n",
    "entities = []\n",
    "for l in train_data[1:]:\n",
    "    if l == '':\n",
    "        train_docs.append(ncbi_doc(title, text, entities))\n",
    "        point = 0\n",
    "        entities = []\n",
    "    else:\n",
    "        if point == 0:\n",
    "            title = l\n",
    "        elif point == 1:\n",
    "            text = l\n",
    "        else:\n",
    "            entities.append(l)\n",
    "        point += 1\n",
    "len(train_docs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "99"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dev_docs = []\n",
    "point = 0\n",
    "entities = []\n",
    "for l in dev_data[1:]:\n",
    "    if l == '':\n",
    "        dev_docs.append(ncbi_doc(title, text, entities))\n",
    "        point = 0\n",
    "        entities = []\n",
    "    else:\n",
    "        if point == 0:\n",
    "            title = l\n",
    "        elif point == 1:\n",
    "            text = l\n",
    "        else:\n",
    "            entities.append(l)\n",
    "        point += 1\n",
    "len(dev_docs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using spacy tokenizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "nlp = spacy.load('en')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('tagger', <spacy.pipeline.Tagger at 0x7f0b714f9080>),\n",
       " ('parser', <spacy.pipeline.DependencyParser at 0x7f0b71531f68>),\n",
       " ('ner', <spacy.pipeline.EntityRecognizer at 0x7f0b71531fc0>)]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# specifying nlp tokenizer to be consistent \n",
    "def create_custom_tokenizer(nlp):\n",
    "    \n",
    "    my_prefix = [r'[0-9]\\.']\n",
    "    \n",
    "    all_prefixes_re = spacy.util.compile_prefix_regex(tuple(list(nlp.Defaults.prefixes) + my_prefix))\n",
    "    \n",
    "    # Handle ( that doesn't have proper spacing around it\n",
    "    custom_infixes = ['\\.\\.\\.+', '(?<=[0-9])-(?=[0-9])','(\\w\\w*)-(\\w\\w*)', '[!&:,()]']\n",
    "    infix_re = spacy.util.compile_infix_regex(tuple(list(nlp.Defaults.infixes) + custom_infixes))\n",
    "    \n",
    "    suffix_re = spacy.util.compile_suffix_regex(nlp.Defaults.suffixes)   \n",
    "    \n",
    "    return Tokenizer(nlp.vocab, nlp.Defaults.tokenizer_exceptions,\n",
    "                     prefix_search = all_prefixes_re.search, \n",
    "                     infix_finditer = infix_re.finditer, suffix_search = suffix_re.search,\n",
    "                     token_match=None)\n",
    "nlp.tokenizer = create_custom_tokenizer(nlp)\n",
    "nlp.pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ratio #dev/(#dev + #train) = 99/691 = 0.143\n"
     ]
    }
   ],
   "source": [
    "print('ratio #dev/(#dev + #train) = {}/{} = {:.3}'.format\\\n",
    "      (len(dev_docs), len(dev_docs) + len(train_docs), len(dev_docs)/(len(dev_docs) + len(train_docs))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write to conll"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/592 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 592 documents. Processing conversion...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 592/592 [00:49<00:00, 11.86it/s]\n",
      "  2%|▏         | 2/99 [00:00<00:06, 14.00it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 99 documents. Processing conversion...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 99/99 [00:07<00:00, 12.97it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['Somatic-cell\\tO\\n',\n",
       " 'selection\\tO\\n',\n",
       " 'is\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'major\\tO\\n',\n",
       " 'determinant\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'blood-cell\\tO\\n',\n",
       " 'phenotype\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'heterozygotes\\tO\\n',\n",
       " 'for\\tO\\n',\n",
       " 'glucose-6-phosphate\\tO\\n',\n",
       " 'dehydrogenase\\tO\\n',\n",
       " 'mutations\\tO\\n',\n",
       " 'causing\\tO\\n',\n",
       " 'severe\\tO\\n',\n",
       " 'enzyme\\tB-DiseaseClass\\n',\n",
       " 'deficiency\\tI-DiseaseClass\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'X-chromosome\\tO\\n',\n",
       " 'inactivation\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'mammals\\tO\\n',\n",
       " 'is\\tO\\n',\n",
       " 'regarded\\tO\\n',\n",
       " 'as\\tO\\n',\n",
       " 'an\\tO\\n',\n",
       " 'essentially\\tO\\n',\n",
       " 'random\\tO\\n',\n",
       " 'process\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'but\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'resulting\\tO\\n',\n",
       " 'somatic-cell\\tO\\n',\n",
       " 'mosaicism\\tO\\n',\n",
       " 'creates\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'opportunity\\tO\\n',\n",
       " 'for\\tO\\n',\n",
       " 'cell\\tO\\n',\n",
       " 'selection\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'In\\tO\\n',\n",
       " 'most\\tO\\n',\n",
       " 'people\\tO\\n',\n",
       " 'with\\tO\\n',\n",
       " 'red-blood\\tO\\n',\n",
       " '-\\tO\\n',\n",
       " 'cell\\tO\\n',\n",
       " 'glucose-6-phosphate\\tB-SpecificDisease\\n',\n",
       " 'dehydrogenase\\tI-SpecificDisease\\n',\n",
       " '(G6PD)\\tI-SpecificDisease\\n',\n",
       " 'G6PD\\tO\\n',\n",
       " ')\\tO\\n',\n",
       " 'deficiency\\tB-SpecificDisease\\n',\n",
       " ',\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'enzyme-deficient\\tO\\n',\n",
       " 'phenotype\\tO\\n',\n",
       " 'is\\tO\\n',\n",
       " 'only\\tO\\n',\n",
       " 'moderately\\tO\\n',\n",
       " 'expressed\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'nucleated\\tO\\n',\n",
       " 'cells\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'However\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'small\\tO\\n',\n",
       " 'subset\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'hemizygous\\tO\\n',\n",
       " 'males\\tO\\n',\n",
       " 'who\\tO\\n',\n",
       " 'suffer\\tO\\n',\n",
       " 'from\\tO\\n',\n",
       " 'chronic\\tB-SpecificDisease\\n',\n",
       " 'nonspherocytic\\tI-SpecificDisease\\n',\n",
       " 'hemolytic\\tI-SpecificDisease\\n',\n",
       " 'anemia\\tI-SpecificDisease\\n',\n",
       " ',\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'underlying\\tO\\n',\n",
       " 'mutations\\tO\\n',\n",
       " '(\\tO\\n',\n",
       " 'designated\\tO\\n',\n",
       " 'class\\tO\\n',\n",
       " 'I\\tO\\n',\n",
       " ')\\tO\\n',\n",
       " 'cause\\tO\\n',\n",
       " 'more-severe\\tO\\n',\n",
       " 'G6PD\\tB-SpecificDisease\\n',\n",
       " 'deficiency\\tI-SpecificDisease\\n',\n",
       " ',\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'this\\tO\\n',\n",
       " 'might\\tO\\n',\n",
       " 'provide\\tO\\n',\n",
       " 'an\\tO\\n',\n",
       " 'opportunity\\tO\\n',\n",
       " 'for\\tO\\n',\n",
       " 'selection\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'heterozygous\\tO\\n',\n",
       " 'females\\tO\\n',\n",
       " 'during\\tO\\n',\n",
       " 'development\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'In\\tO\\n',\n",
       " 'order\\tO\\n',\n",
       " 'to\\tO\\n',\n",
       " 'test\\tO\\n',\n",
       " 'this\\tO\\n',\n",
       " 'possibility\\tO\\n',\n",
       " 'we\\tO\\n',\n",
       " 'have\\tO\\n',\n",
       " 'analyzed\\tO\\n',\n",
       " 'four\\tO\\n',\n",
       " 'heterozygotes\\tO\\n',\n",
       " 'for\\tO\\n',\n",
       " 'class\\tO\\n',\n",
       " '\\n',\n",
       " 'I\\tO\\n',\n",
       " 'G6PD\\tO\\n',\n",
       " 'mutations\\tO\\n',\n",
       " 'two\\tO\\n',\n",
       " 'with\\tO\\n',\n",
       " 'G6PD\\tO\\n',\n",
       " 'Portici\\tO\\n',\n",
       " '(\\tO\\n',\n",
       " '1178G--\\tO\\n',\n",
       " '>\\tO\\n',\n",
       " '\\n',\n",
       " 'A\\tO\\n',\n",
       " ')\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'two\\tO\\n',\n",
       " 'with\\tO\\n',\n",
       " 'G6PD\\tO\\n',\n",
       " 'Bari\\tO\\n',\n",
       " '(\\tO\\n',\n",
       " '1187C--\\tO\\n',\n",
       " '>\\tO\\n',\n",
       " 'T\\tO\\n',\n",
       " ')\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'We\\tO\\n',\n",
       " 'found\\tO\\n',\n",
       " 'that\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'fractionated\\tO\\n',\n",
       " 'blood\\tO\\n',\n",
       " 'cell\\tO\\n',\n",
       " 'types\\tO\\n',\n",
       " '(\\tO\\n',\n",
       " 'including\\tO\\n',\n",
       " 'erythroid\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'myeloid\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'lymphoid\\tO\\n',\n",
       " 'cell\\tO\\n',\n",
       " 'lineages\\tO\\n',\n",
       " ')\\tO\\n',\n",
       " 'there\\tO\\n',\n",
       " 'was\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'significant\\tO\\n',\n",
       " 'excess\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'G6PD-normal\\tO\\n',\n",
       " 'cells\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'The\\tO\\n',\n",
       " 'significant\\tO\\n',\n",
       " 'concordance\\tO\\n',\n",
       " 'that\\tO\\n',\n",
       " 'we\\tO\\n',\n",
       " 'have\\tO\\n',\n",
       " 'observed\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'degree\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'imbalance\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'different\\tO\\n',\n",
       " 'blood-cell\\tO\\n',\n",
       " 'lineages\\tO\\n',\n",
       " 'indicates\\tO\\n',\n",
       " 'that\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'selective\\tO\\n',\n",
       " 'mechanism\\tO\\n',\n",
       " 'is\\tO\\n',\n",
       " 'likely\\tO\\n',\n",
       " 'to\\tO\\n',\n",
       " 'operate\\tO\\n',\n",
       " 'at\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'level\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'pluripotent\\tO\\n',\n",
       " 'blood\\tO\\n',\n",
       " 'stem\\tO\\n',\n",
       " 'cells\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'Thus\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'it\\tO\\n',\n",
       " 'appears\\tO\\n',\n",
       " 'that\\tO\\n',\n",
       " 'severe\\tO\\n',\n",
       " 'G6PD\\tB-SpecificDisease\\n',\n",
       " 'deficiency\\tI-SpecificDisease\\n',\n",
       " 'affects\\tO\\n',\n",
       " 'adversely\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'proliferation\\tO\\n',\n",
       " 'or\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'survival\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'nucleated\\tO\\n',\n",
       " 'blood\\tO\\n',\n",
       " 'cells\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'that\\tO\\n',\n",
       " 'this\\tO\\n',\n",
       " 'phenotypic\\tO\\n',\n",
       " 'characteristic\\tO\\n',\n",
       " 'is\\tO\\n',\n",
       " 'critical\\tO\\n',\n",
       " 'during\\tO\\n',\n",
       " 'hematopoiesis\\tO\\n',\n",
       " '..\\tO\\n',\n",
       " '\\n',\n",
       " '\\n',\n",
       " 'The\\tO\\n',\n",
       " 'ataxia-telangiectasia\\tB-Modifier\\n',\n",
       " 'gene\\tO\\n',\n",
       " 'product\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'constitutively\\tO\\n',\n",
       " 'expressed\\tO\\n',\n",
       " 'nuclear\\tO\\n',\n",
       " 'protein\\tO\\n',\n",
       " 'that\\tO\\n',\n",
       " 'is\\tO\\n',\n",
       " 'not\\tO\\n',\n",
       " 'up-regulated\\tO\\n',\n",
       " 'following\\tO\\n',\n",
       " 'genome\\tO\\n',\n",
       " 'damage\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'The\\tO\\n',\n",
       " 'product\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'ataxia-telangiectasia\\tB-Modifier\\n',\n",
       " 'gene\\tO\\n',\n",
       " '(\\tO\\n',\n",
       " 'ATM\\tO\\n',\n",
       " ')\\tO\\n',\n",
       " 'was\\tO\\n',\n",
       " 'identified\\tO\\n',\n",
       " 'by\\tO\\n',\n",
       " 'using\\tO\\n',\n",
       " 'an\\tO\\n',\n",
       " 'antiserum\\tO\\n',\n",
       " 'developed\\tO\\n',\n",
       " 'to\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'peptide\\tO\\n',\n",
       " 'corresponding\\tO\\n',\n",
       " 'to\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'deduced\\tO\\n',\n",
       " 'amino\\tO\\n',\n",
       " 'acid\\tO\\n',\n",
       " 'sequence\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'The\\tO\\n',\n",
       " 'ATM\\tO\\n',\n",
       " 'protein\\tO\\n',\n",
       " 'is\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'single\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'high-molecular\\tO\\n',\n",
       " 'weight\\tO\\n',\n",
       " 'protein\\tO\\n',\n",
       " 'predominantly\\tO\\n',\n",
       " 'confined\\tO\\n',\n",
       " 'to\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'nucleus\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'human\\tO\\n',\n",
       " 'fibroblasts\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'but\\tO\\n',\n",
       " 'is\\tO\\n',\n",
       " 'present\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'both\\tO\\n',\n",
       " 'nuclear\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'microsomal\\tO\\n',\n",
       " 'fractions\\tO\\n',\n",
       " 'from\\tO\\n',\n",
       " 'human\\tO\\n',\n",
       " 'lymphoblast\\tO\\n',\n",
       " 'cells\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'peripheral\\tO\\n',\n",
       " 'blood\\tO\\n',\n",
       " 'lymphocytes\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'ATM\\tO\\n',\n",
       " 'protein\\tO\\n',\n",
       " 'levels\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'localization\\tO\\n',\n",
       " 'remain\\tO\\n',\n",
       " 'constant\\tO\\n',\n",
       " 'throughout\\tO\\n',\n",
       " 'all\\tO\\n',\n",
       " 'stages\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'cell\\tO\\n',\n",
       " 'cycle\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'Truncated\\tO\\n',\n",
       " 'ATM\\tO\\n',\n",
       " 'protein\\tO\\n',\n",
       " 'was\\tO\\n',\n",
       " 'not\\tO\\n',\n",
       " 'detected\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'lymphoblasts\\tO\\n',\n",
       " 'from\\tO\\n',\n",
       " 'ataxia-telangiectasia\\tB-Modifier\\n',\n",
       " 'patients\\tO\\n',\n",
       " 'homozygous\\tO\\n',\n",
       " 'for\\tO\\n',\n",
       " 'mutations\\tO\\n',\n",
       " 'leading\\tO\\n',\n",
       " 'to\\tO\\n',\n",
       " 'premature\\tO\\n',\n",
       " 'protein\\tO\\n',\n",
       " 'termination\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'Exposure\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'normal\\tO\\n',\n",
       " 'human\\tO\\n',\n",
       " 'cells\\tO\\n',\n",
       " 'to\\tO\\n',\n",
       " 'gamma-irradiation\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'radiomimetic\\tO\\n',\n",
       " 'drug\\tO\\n',\n",
       " 'neocarzinostatin\\tO\\n',\n",
       " 'had\\tO\\n',\n",
       " 'no\\tO\\n',\n",
       " 'effect\\tO\\n',\n",
       " 'on\\tO\\n',\n",
       " 'ATM\\tO\\n',\n",
       " 'protein\\tO\\n',\n",
       " 'levels\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'contrast\\tO\\n',\n",
       " 'to\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'noted\\tO\\n',\n",
       " 'rise\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'p53\\tO\\n',\n",
       " 'levels\\tO\\n',\n",
       " 'over\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'same\\tO\\n',\n",
       " 'time\\tO\\n',\n",
       " 'interval\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'These\\tO\\n',\n",
       " 'findings\\tO\\n',\n",
       " 'are\\tO\\n',\n",
       " 'consistent\\tO\\n',\n",
       " 'with\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'role\\tO\\n',\n",
       " 'for\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'ATM\\tO\\n',\n",
       " 'protein\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'ensuring\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'fidelity\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'DNA\\tO\\n',\n",
       " 'repair\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'cell\\tO\\n',\n",
       " 'cycle\\tO\\n',\n",
       " 'regulation\\tO\\n',\n",
       " 'following\\tO\\n',\n",
       " 'genome\\tO\\n',\n",
       " 'damage\\tO\\n',\n",
       " '..\\tO\\n',\n",
       " '\\n',\n",
       " '\\n',\n",
       " 'Molecular\\tO\\n',\n",
       " 'basis\\tO\\n',\n",
       " 'for\\tO\\n',\n",
       " 'Duarte\\tB-CompositeMention\\n',\n",
       " 'and\\tI-CompositeMention\\n',\n",
       " 'Los\\tI-CompositeMention\\n',\n",
       " 'Angeles\\tI-CompositeMention\\n',\n",
       " 'variant\\tI-CompositeMention\\n',\n",
       " 'galactosemia\\tI-CompositeMention\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'Human\\tO\\n',\n",
       " 'orythrocytes\\tO\\n',\n",
       " 'that\\tO\\n',\n",
       " 'are\\tO\\n',\n",
       " 'homozygous\\tO\\n',\n",
       " 'for\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'Duarte\\tB-SpecificDisease\\n',\n",
       " 'enzyme\\tI-SpecificDisease\\n',\n",
       " 'variant\\tI-SpecificDisease\\n',\n",
       " 'of\\tI-SpecificDisease\\n',\n",
       " 'galactosemia\\tI-SpecificDisease\\n',\n",
       " '(\\tO\\n',\n",
       " 'D\\tO\\n',\n",
       " '/\\tO\\n',\n",
       " 'D\\tO\\n',\n",
       " ')\\tO\\n',\n",
       " 'have\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'characteristic\\tO\\n',\n",
       " 'isoform\\tO\\n',\n",
       " 'on\\tO\\n',\n",
       " 'isoelectric\\tO\\n',\n",
       " 'focusing\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " '50\\tO\\n',\n",
       " '%\\tO\\n',\n",
       " 'reduction\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'galactose-1-phosphate\\tO\\n',\n",
       " 'uridyltransferase\\tO\\n',\n",
       " '(\\tO\\n',\n",
       " 'GALT\\tO\\n',\n",
       " ')\\tO\\n',\n",
       " 'enzyme\\tO\\n',\n",
       " 'activity\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'The\\tO\\n',\n",
       " 'Duarte\\tO\\n',\n",
       " 'biochemical\\tO\\n',\n",
       " 'phenotype\\tO\\n',\n",
       " 'has\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'molecular\\tO\\n',\n",
       " 'genotype\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'N314D\\tO\\n',\n",
       " '/\\tO\\n',\n",
       " 'N314D.\\tO\\n',\n",
       " '\\n',\n",
       " 'The\\tO\\n',\n",
       " 'characteristic\\tO\\n',\n",
       " 'Duarte\\tO\\n',\n",
       " 'isoform\\tO\\n',\n",
       " 'is\\tO\\n',\n",
       " 'also\\tO\\n',\n",
       " 'associated\\tO\\n',\n",
       " 'with\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'variant\\tO\\n',\n",
       " 'called\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " '\"\\tO\\n',\n",
       " 'Los\\tO\\n',\n",
       " 'Angeles\\tO\\n',\n",
       " '(\\tO\\n',\n",
       " 'LA\\tO\\n',\n",
       " ')\\tO\\n',\n",
       " 'phenotype\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " '\"\\tO\\n',\n",
       " 'which\\tO\\n',\n",
       " 'has\\tO\\n',\n",
       " 'increased\\tO\\n',\n",
       " 'GALT\\tO\\n',\n",
       " 'enzyme\\tO\\n',\n",
       " 'activity\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'We\\tO\\n',\n",
       " 'evaluated\\tO\\n',\n",
       " 'GALT\\tO\\n',\n",
       " 'enzyme\\tO\\n',\n",
       " 'activity\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'screened\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'GALT\\tO\\n',\n",
       " 'genes\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " '145\\tO\\n',\n",
       " 'patients\\tO\\n',\n",
       " 'with\\tO\\n',\n",
       " 'one\\tO\\n',\n",
       " 'or\\tO\\n',\n",
       " 'more\\tO\\n',\n",
       " 'N314D-containing\\tO\\n',\n",
       " 'alleles\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'We\\tO\\n',\n",
       " 'found\\tO\\n',\n",
       " 'seven\\tO\\n',\n",
       " 'with\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'LA\\tO\\n',\n",
       " 'biochemical\\tO\\n',\n",
       " 'phenotype\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'all\\tO\\n',\n",
       " 'had\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " '1721C--\\tO\\n',\n",
       " '>\\tO\\n',\n",
       " 'T\\tO\\n',\n",
       " 'transition\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'exon\\tO\\n',\n",
       " '7\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'cis\\tO\\n',\n",
       " 'with\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'N314D\\tO\\n',\n",
       " 'missense\\tO\\n',\n",
       " 'mutation\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'The\\tO\\n',\n",
       " '1721C--\\tO\\n',\n",
       " '>\\tO\\n',\n",
       " 'T\\tO\\n',\n",
       " 'transition\\tO\\n',\n",
       " 'is\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'neutral\\tO\\n',\n",
       " 'polymorphism\\tO\\n',\n",
       " 'for\\tO\\n',\n",
       " 'leucine\\tO\\n',\n",
       " 'at\\tO\\n',\n",
       " 'amino\\tO\\n',\n",
       " 'acid\\tO\\n',\n",
       " '218\\tO\\n',\n",
       " '(\\tO\\n',\n",
       " 'L218L\\tO\\n',\n",
       " ')\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'In\\tO\\n',\n",
       " 'pedigree\\tO\\n',\n",
       " 'analyses\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'this\\tO\\n',\n",
       " '1721C--\\tO\\n',\n",
       " '>\\tO\\n',\n",
       " 'T\\tO\\n',\n",
       " 'transition\\tO\\n',\n",
       " 'segregated\\tO\\n',\n",
       " 'with\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'LA\\tO\\n',\n",
       " 'phenotype\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'increased\\tO\\n',\n",
       " 'GALT\\tO\\n',\n",
       " 'activity\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'three\\tO\\n',\n",
       " 'different\\tO\\n',\n",
       " 'biochemical\\tO\\n',\n",
       " 'phenotypes\\tO\\n',\n",
       " '(\\tO\\n',\n",
       " 'LA\\tO\\n',\n",
       " '/\\tO\\n',\n",
       " 'N\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'LA\\tO\\n',\n",
       " '/\\tO\\n',\n",
       " 'G\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'LA\\tO\\n',\n",
       " '/\\tO\\n',\n",
       " 'D\\tO\\n',\n",
       " ')\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'To\\tO\\n',\n",
       " 'determine\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'mechanism\\tO\\n',\n",
       " 'for\\tO\\n',\n",
       " 'increased\\tO\\n',\n",
       " 'activity\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'LA\\tO\\n',\n",
       " 'variant\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'we\\tO\\n',\n",
       " 'compared\\tO\\n',\n",
       " 'GALT\\tO\\n',\n",
       " 'mRNA\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'protein\\tO\\n',\n",
       " 'abundance\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'enzyme\\tO\\n',\n",
       " 'thermal\\tO\\n',\n",
       " 'stability\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'lymphoblast\\tO\\n',\n",
       " 'cell\\tO\\n',\n",
       " 'lines\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'D\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'LA\\tO\\n',\n",
       " 'phenotypes\\tO\\n',\n",
       " 'with\\tO\\n',\n",
       " 'comparable\\tO\\n',\n",
       " 'genotypes\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'GALT\\tO\\n',\n",
       " 'protein\\tO\\n',\n",
       " 'abundance\\tO\\n',\n",
       " 'was\\tO\\n',\n",
       " 'increased\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'LA\\tO\\n',\n",
       " 'compared\\tO\\n',\n",
       " 'to\\tO\\n',\n",
       " 'D\\tO\\n',\n",
       " 'alleles\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'but\\tO\\n',\n",
       " 'mRNA\\tO\\n',\n",
       " 'was\\tO\\n',\n",
       " 'similar\\tO\\n',\n",
       " 'among\\tO\\n',\n",
       " 'all\\tO\\n',\n",
       " 'genotypes\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'When\\tO\\n',\n",
       " 'LA\\tO\\n',\n",
       " '/\\tO\\n',\n",
       " 'D\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'D\\tO\\n',\n",
       " '/\\tO\\n',\n",
       " 'D\\tO\\n',\n",
       " 'GALT\\tO\\n',\n",
       " 'biochemical\\tO\\n',\n",
       " 'phenotypes\\tO\\n',\n",
       " 'were\\tO\\n',\n",
       " 'compared\\tO\\n',\n",
       " 'to\\tO\\n',\n",
       " 'N\\tO\\n',\n",
       " '/\\tO\\n',\n",
       " 'N\\tO\\n',\n",
       " 'GALT\\tO\\n',\n",
       " 'phenotypes\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'both\\tO\\n',\n",
       " 'had\\tO\\n',\n",
       " '50\\tO\\n',\n",
       " '%\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'as\\tO\\n',\n",
       " 'compared\\tO\\n',\n",
       " 'to\\tO\\n',\n",
       " '21\\tO\\n',\n",
       " '%\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'reduction\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'GALT\\tO\\n',\n",
       " 'activity\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'wild\\tO\\n',\n",
       " 'type\\tO\\n',\n",
       " '(\\tO\\n',\n",
       " 'N\\tO\\n',\n",
       " '/\\tO\\n',\n",
       " 'N\\tO\\n',\n",
       " ')\\tO\\n',\n",
       " 'after\\tO\\n',\n",
       " 'exposure\\tO\\n',\n",
       " 'at\\tO\\n',\n",
       " 'identical\\tO\\n',\n",
       " 'initial\\tO\\n',\n",
       " 'enzyme\\tO\\n',\n",
       " 'activity\\tO\\n',\n",
       " 'to\\tO\\n',\n",
       " '50\\tO\\n',\n",
       " 'degrees\\tO\\n',\n",
       " 'C\\tO\\n',\n",
       " 'for\\tO\\n',\n",
       " '15\\tO\\n',\n",
       " 'min\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'We\\tO\\n',\n",
       " 'conclude\\tO\\n',\n",
       " 'that\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'codon\\tO\\n',\n",
       " 'change\\tO\\n',\n",
       " 'N314D\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'cis\\tO\\n',\n",
       " 'with\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'base-pair\\tO\\n',\n",
       " 'transition\\tO\\n',\n",
       " '1721C--\\tO\\n',\n",
       " '>\\tO\\n',\n",
       " '\\n',\n",
       " 'T\\tO\\n',\n",
       " 'produces\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'LA\\tB-SpecificDisease\\n',\n",
       " 'variant\\tI-SpecificDisease\\n',\n",
       " 'of\\tI-SpecificDisease\\n',\n",
       " 'galactosemia\\tI-SpecificDisease\\n',\n",
       " 'and\\tO\\n',\n",
       " 'that\\tO\\n',\n",
       " 'this\\tO\\n',\n",
       " 'nucleotide\\tO\\n',\n",
       " 'change\\tO\\n',\n",
       " 'increases\\tO\\n',\n",
       " 'GALT\\tO\\n',\n",
       " 'activity\\tO\\n',\n",
       " 'by\\tO\\n',\n",
       " 'increasing\\tO\\n',\n",
       " 'GALT\\tO\\n',\n",
       " 'protein\\tO\\n',\n",
       " 'abundance\\tO\\n',\n",
       " 'without\\tO\\n',\n",
       " 'increasing\\tO\\n',\n",
       " 'transcription\\tO\\n',\n",
       " 'or\\tO\\n',\n",
       " 'decreasing\\tO\\n',\n",
       " 'thermal\\tO\\n',\n",
       " 'lability\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'A\\tO\\n',\n",
       " 'favorable\\tO\\n',\n",
       " 'codon\\tO\\n',\n",
       " 'bias\\tO\\n',\n",
       " 'for\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'mutated\\tO\\n',\n",
       " 'codon\\tO\\n',\n",
       " 'with\\tO\\n',\n",
       " 'consequently\\tO\\n',\n",
       " 'increased\\tO\\n',\n",
       " 'translation\\tO\\n',\n",
       " 'rates\\tO\\n',\n",
       " 'is\\tO\\n',\n",
       " 'postulated\\tO\\n',\n",
       " 'as\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'mechanism\\tO\\n',\n",
       " '..\\tO\\n',\n",
       " '\\n',\n",
       " '\\n',\n",
       " 'An\\tO\\n',\n",
       " 'intronic\\tO\\n',\n",
       " 'mutation\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'lariat\\tO\\n',\n",
       " 'branchpoint\\tO\\n',\n",
       " 'sequence\\tO\\n',\n",
       " 'is\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'direct\\tO\\n',\n",
       " 'cause\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'an\\tO\\n',\n",
       " 'inherited\\tB-DiseaseClass\\n',\n",
       " 'human\\tI-DiseaseClass\\n',\n",
       " 'disorder\\tI-DiseaseClass\\n',\n",
       " '(\\tO\\n',\n",
       " 'fish-eye\\tB-SpecificDisease\\n',\n",
       " 'disease\\tI-SpecificDisease\\n',\n",
       " ')\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'The\\tO\\n',\n",
       " 'first\\tO\\n',\n",
       " 'step\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'splicing\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'an\\tO\\n',\n",
       " 'intron\\tO\\n',\n",
       " 'from\\tO\\n',\n",
       " 'nuclear\\tO\\n',\n",
       " 'precursors\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'mRNA\\tO\\n',\n",
       " 'results\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'formation\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'lariat\\tO\\n',\n",
       " 'structure\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'A\\tO\\n',\n",
       " 'distinct\\tO\\n',\n",
       " 'intronic\\tO\\n',\n",
       " 'nucleotide\\tO\\n',\n",
       " 'sequence\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'known\\tO\\n',\n",
       " 'as\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'branchpoint\\tO\\n',\n",
       " 'region\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'plays\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'central\\tO\\n',\n",
       " 'role\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'this\\tO\\n',\n",
       " 'process\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'We\\tO\\n',\n",
       " 'here\\tO\\n',\n",
       " 'describe\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'point\\tO\\n',\n",
       " 'mutation\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'such\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'sequence\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'Three\\tO\\n',\n",
       " 'sisters\\tO\\n',\n",
       " 'were\\tO\\n',\n",
       " 'shown\\tO\\n',\n",
       " 'to\\tO\\n',\n",
       " 'suffer\\tO\\n',\n",
       " 'from\\tO\\n',\n",
       " 'fish-eye\\tB-SpecificDisease\\n',\n",
       " 'disease\\tI-SpecificDisease\\n',\n",
       " '(\\tO\\n',\n",
       " 'FED\\tB-SpecificDisease\\n',\n",
       " ')\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'disorder\\tO\\n',\n",
       " 'which\\tO\\n',\n",
       " 'is\\tO\\n',\n",
       " 'caused\\tO\\n',\n",
       " 'by\\tO\\n',\n",
       " 'mutations\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'gene\\tO\\n',\n",
       " 'coding\\tO\\n',\n",
       " 'for\\tO\\n',\n",
       " 'lecithin\\tO\\n',\n",
       " 'cholesterol\\tO\\n',\n",
       " 'acyltransferase\\tO\\n',\n",
       " '(\\tO\\n',\n",
       " 'LCAT\\tO\\n',\n",
       " ')\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'Sequencing\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'LCAT\\tO\\n',\n",
       " 'gene\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'all\\tO\\n',\n",
       " 'three\\tO\\n',\n",
       " 'probands\\tO\\n',\n",
       " 'revealed\\tO\\n',\n",
       " 'compound\\tO\\n',\n",
       " 'heterozygosity\\tO\\n',\n",
       " 'for\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'missense\\tO\\n',\n",
       " 'mutation\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'exon\\tO\\n',\n",
       " '4\\tO\\n',\n",
       " 'which\\tO\\n',\n",
       " 'is\\tO\\n',\n",
       " 'reported\\tO\\n',\n",
       " 'to\\tO\\n',\n",
       " 'underlie\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'FED\\tB-Modifier\\n',\n",
       " 'phenotype\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'a\\tO\\n',\n",
       " 'point\\tO\\n',\n",
       " 'mutation\\tO\\n',\n",
       " 'located\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'intron\\tO\\n',\n",
       " '4\\tO\\n',\n",
       " '(\\tO\\n',\n",
       " 'IVS4\\tO\\n',\n",
       " 'T-22C\\tO\\n',\n",
       " ')\\tO\\n',\n",
       " '.\\tO\\n',\n",
       " '\\n',\n",
       " 'By\\tO\\n',\n",
       " 'performing\\tO\\n',\n",
       " 'in\\tO\\n',\n",
       " 'vitro\\tO\\n',\n",
       " 'expression\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'LCAT\\tO\\n',\n",
       " 'minigenes\\tO\\n',\n",
       " 'and\\tO\\n',\n",
       " 'reverse\\tO\\n',\n",
       " 'transcriptase\\tO\\n',\n",
       " 'PCR\\tO\\n',\n",
       " 'on\\tO\\n',\n",
       " 'mRNA\\tO\\n',\n",
       " 'isolated\\tO\\n',\n",
       " 'from\\tO\\n',\n",
       " 'leukocytes\\tO\\n',\n",
       " 'of\\tO\\n',\n",
       " 'the\\tO\\n',\n",
       " 'patient\\tO\\n',\n",
       " ',\\tO\\n',\n",
       " ...]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myCorpus_brat2conll(train_docs, nlp, path2train_write)\n",
    "myCorpus_brat2conll(dev_docs, nlp, path2dev_write)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Corpus containing 128557 tokens in 6233 sentences\n",
      "Corpus containing 22508 tokens in 1046 sentences\n",
      "O : 139715\n",
      "DiseaseClass : 1955\n",
      "SpecificDisease : 6595\n",
      "CompositeMention : 717\n",
      "Modifier : 2083\n",
      "# of annotations 11350\n"
     ]
    }
   ],
   "source": [
    "train_sum = describe_entities('../data/ncbi_train.conll')\n",
    "dev_sum = describe_entities('../data/ncbi_dev.conll')\n",
    "\n",
    "all_ann = 0\n",
    "for ent in train_sum.keys():\n",
    "    print('{} : {}'.format(ent, train_sum[ent] + dev_sum[ent]))\n",
    "    if ent != 'O':\n",
    "        all_ann += train_sum[ent] + dev_sum[ent]\n",
    "print('# of annotations {}'.format(all_ann))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Verifying that there are no missing entities after conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "path2conll = '../data/ncbi_train.conll'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'DiseaseClass': 1708,\n",
       "         'SpecificDisease': 5756,\n",
       "         'CompositeMention': 558,\n",
       "         'Modifier': 1755})"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "brat_ents = []\n",
    "for doc in train_docs:\n",
    "    for ent in doc.entities:\n",
    "        for w in ent.word.split(' '):\n",
    "            brat_ents.append(ent.label)\n",
    "Counter(brat_ents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Corpus containing 128557 tokens in 6233 sentences with 9775 non-O tags\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Counter({'O': 118782,\n",
       "         'DiseaseClass': 1708,\n",
       "         'SpecificDisease': 5755,\n",
       "         'CompositeMention': 557,\n",
       "         'Modifier': 1755})"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "describe_entities(path2conll)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'DiseaseClass': 247,\n",
       "         'SpecificDisease': 841,\n",
       "         'Modifier': 328,\n",
       "         'CompositeMention': 160})"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "brat_ents = []\n",
    "for doc in dev_docs:\n",
    "    for ent in doc.entities:\n",
    "        for w in ent.word.split(' '):\n",
    "            brat_ents.append(ent.label)\n",
    "Counter(brat_ents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Corpus containing 22508 tokens in 1046 sentences with 1575 non-O tags\n",
      "Counter({'O': 20933, 'SpecificDisease': 840, 'Modifier': 328, 'DiseaseClass': 247, 'CompositeMention': 160})\n",
      "0.07516556291390729\n",
      "11350\n"
     ]
    }
   ],
   "source": [
    "path2dev = '../data/ncbi_dev.conll'\n",
    "print(describe_entities(path2dev))\n",
    "tot_tagged = 9775 + 1575\n",
    "print(tot_tagged/151000)\n",
    "print(tot_tagged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
